core/src/main/java/io/grpc/inprocess/InProcessServer.java

```diff
@@ -71,15 +81,17 @@ static InProcessServer findServer(String name) {
   public void start(ServerListener serverListener) throws IOException {
     this.listener = serverListener;
     this.scheduler = schedulerPool.getObject();
-    // Must be last, as channels can start connecting after this point.
-    if (registry.putIfAbsent(name, this) != null) {
-      throw new IOException("name already registered: " + name);
+    if (!anonymous) {
+      // Must be last, as channels can start connecting after this point.
+      if (registry.putIfAbsent(name, this) != null) {
+        throw new IOException("name already registered: " + name);
+      }
     }
   }
 
   @Override
   public SocketAddress getListenSocketAddress() {
-    return new InProcessSocketAddress(name);
+    return new InProcessSocketAddress(name, anonymous ? this : null);
```

Comment:
This choice is notable. The address instance will only reference the server if it's anonymous.

Reply:
Yes, I think that is the correct behavior.

At first I was like, "oh, definitely," then I realized "but that might produce weird behavior/semantics" because depending on how the address was created you'd get different behavior. But I actually think it works well. Essentially it is the same as the anonymous server having a randomly-generated ID that is never reused.